### DISCUSSION TITLE
Safe Exam Browser 3.0

### DISCUSSION THREAD

**Discussion by danschlet** [SafeExamBrowser/seb-mac#139] [General]
SEB 3.0 for macOS is a major update, supporting the modern WebKit browser engine and access to camera/microphone via WebRTC.

- Support for the modern WebKit browser engine (WKWebView), which increases performance, stability and compatibility with web applications. 

- Access to camera/microphone (_getUserMedia_ Web API) via WebRTC on macOS 11.1 Big Sur and newer. You can use web-based video conferencing and remote proctoring solutions, as long as the requirements explained below are met, so that SEB can use the modern WebView. Note: Apple currently doesn't support _getDisplayMedia_ in WKWebView for screen sharing.

- Due to refactored code using newer APIs, SEB 3.0 requires macOS 10.11 or later. We recommend to only use macOS versions which still receive security updates from Apple (the three most recent, currently macOS 10.14, 10.15 and 11).

- When using the modern WebKit browser engine (WKWebView), SEB currently doesn't support _URL content filters_ (only URL filters for links and page loading) and _sending the Config Key and Browser Exam Key in HTTP headers_ (the latter feature will not be possible with WKWebView at all, so you will need to transition to the SEB JavaScript API in your web applications, see the [developer documentation](https://safeexambrowser.org/developer/seb-config-key.html)). If one of the settings _sendBrowserExamKey=true_ or _URLFilterEnableContentFilter=true_ is enabled, the classic WebView is used depending on the new setting _Preferences/ Browser/"Select Browser Engine Policy"_ (key _browserWindowWebView_), _"Prefer Modern in New Tab+Different Host"_ (value 2), ensures that you still can use testing systems/LMS like Moodle which use the Browser Exam/Config Key test for SEB connecting to the exam (as they open in the classic WebView), but webpages opened in new tabs, which have another host than the exam system, open in WKWebView.

- Support for the new [SEB JavaScript API](https://safeexambrowser.org/developer/seb-config-key.html) for easier integration into exam solutions (currently for querying app version details, ConfigKey and BrowserExamKey security elements).

- Support for restart session ([Preferences/Exam/Link to quit SEB after exam/Restart instead of quitting](https://safeexambrowser.org/macosx/mac_usermanual_en.html#link-to-quit-seb-after-exam)) and quit SEB/session features, which are mostly interesting for managed Macs and kiosk environments:  If client settings are secure (have a quit password set), SEB was not started with a seb(s) link or .seb file and an exam session is quit, SEB is not quit (terminated), but only the exam session (SEB is restarted with client settings).

- Fixes security issue with "Revert to Client Settings" and "Configure Client" in Preferences window.

- Private Clipboard is also supported when using the modern WebView (WKWebView) and working with commands in menu bar / Edit. 

- Added setting in _Preferences / Security / Check full macOS version number_ (key _allowMacOSVersionNumberCheckFull_) for full control about minimal allowed macOS version (major/minor/patch version numbers, keys allowMacOSVersionNumberMajor/ allowMacOSVersionNumberMinor/ allowMacOSVersionNumberPatch).

- Fixed issue with setting _Preferences / Security / Enforce minimal macOS version_.

- Again asking for Admin password when settings aren't encrypted.

- Use AAC kiosk mode only on macOS 15.4+ and 11.4+ (except 11.0 - 11.3, as there was an issue with DNS lookup. This caused pages with yet unknown domains to not load).

- No longer using AAC modal alert workaround on macOS 12 Monterey (as an issue causing application modal alerts to be invisible was fixed there).

- Fixed repeated drag-and-drop onto Preferences window. Note: files in iCloud drive need to be downloaded first in Finder (cloud icon in list view) before they can be dragged-dropped.

- Fixed double toolbar icon for "General" pane when opening settings when the Preferences window already was open.

- Fixed check if reconfiguring is allowed when a .seb config file is opened while SEB is running already (considering the Exam / Allow Reconfiguring setting).

- Now also supporting gzip compressed, unencrypted .seb files.

- Updated _Network / Filter / Teach URL Filter_ feature for modern WebView (WKWebView). Note: If content filtering is enabled, classic WebView is used.

- Fixed Teach URL Filter feature on macOS 11+ also for classic WebView.

- [Setting to enable WebInspector](https://safeexambrowser.org/macosx/mac_usermanual_en.html#BrowserPane) (developer tools) and remote inspection (only when running a debug build) is now functional in WKWebView, although the "Inspect Element" context menu ([right mouse click must be enabled in settings](https://safeexambrowser.org/macosx/mac_usermanual_en.html#SecurityPane) as well) doesn't work correctly with WKWebView.

<hr /><em>This discussion was created from the release <a href='https://github.com/SafeExamBrowser/seb-mac/releases/tag/3.0'>Safe Exam Browser 3.0</a>.</em>

**Comment by mkendrick** [comment]
great looking update!  are 'permitted processes' still unsupported in this version?

**Reply by danschlet** [reply]
Thank you.
See https://github.com/SafeExamBrowser/seb-mac/discussions/114#discussioncomment-1221584

**Reply by danschlet** [reply]
And no, if it would be supported, it would be mentioned in the release notes.

**Reply by mkendrick** [reply]
no worries, thanks for the reply. :)

**Comment by tarikkeskin** [comment]
Greetings. I wonder, whether ı can enable screen sharing on macOS or not. I have read the notes and ı can't completely understand the " while still blocking VNC-based macOS screen sharing and remote management on Mac SEB clients.  " part.  What does that mean? What ı want is pass the screen sharing test, ı dont want to share screen. Thank you.

**Reply by danschlet** [reply]
I'm not sure if that is what you're asking for, but SEB for macOS currently doesn't support web-based screen sharing. The embedded web browser engine (WKWebView) doesn't support the getDisplayMedia() web API, therefore you cannot use for example web-based remote proctoring solutions which try to record/share the screen content. There's a feature request pending for this at Apple.

**Reply by tarikkeskin** [reply]
Alright, thanks for your response

**Comment by tarikkeskin** [comment]
Hello again. I am going to ask one more thing, ı just want to be completely sure. There is a website and it checks mic&camera and Screen sharing. I am passing mic&camera but I also want to pass stream sharing part somehow. Is there any way to do this ?

**Reply by danschlet** [reply]
No, with current SEB versions for macOS it's not possible.
